goshopify.go

```diff
@@ -308,17 +297,35 @@ func (c *Client) Do(req *http.Request, v interface{}) error {
 
 // doGetHeaders executes a request, decoding the response into `v` and also returns any response headers.
 func (c *Client) doGetHeaders(req *http.Request, v interface{}) (http.Header, error) {
-	resp, err := c.Client.Do(req)
-	if err != nil {
-		return nil, err
-	}
-	defer resp.Body.Close()
+	var resp *http.Response
+	var err error
+	for {
+		resp, err = c.Client.Do(req)
+		if err != nil {
+			return nil, err
+		}
 
-	err = CheckResponseError(resp)
-	if err != nil {
-		return nil, err
+		err = CheckResponseError(resp)
+		if err != nil {
+			if retry, isRetryErr := err.(RateLimitError); isRetryErr {
+				// back off and retry
+				if c.retries <= 0 {
+					return nil, err
+				}
+				wait := time.Duration(retry.RetryAfter) * time.Second
+				time.Sleep(wait)
+				c.retries--
```

Comment:
Using the client struct variable `c.retries` means you only have a set number of retries for the life-time of the client and expose yourself to race conditions.
I'd think you want to use a variable local to this function, init it with c.retries and decrease (and check in line 312) that one instead.

Reply:
ya im ok with that